// ucn-community-frontend/src/types/index.ts
export interface User {
  id: string;
  _id?: string;
  name: string;
  email: string;
  carrera: string;
  año_ingreso: number;
  campus: string;
  biografia?: string;
  foto_perfil?: string;
  role: 'student' | 'professor' | 'admin';
  verificado: boolean;
  fecha_registro: Date;
  last_login?: Date;
}

export interface AuthResponse {
  success: boolean;
  message: string;
  data: {
    user: User;
    accessToken: string;
    refreshToken: string;
    tokenExpiry: string;
    requiresEmailVerification?: boolean;
  };
}

export interface LoginData {
  email: string;
  password: string;
  remember_me?: boolean;
}

export interface RegisterData {
  nombre: string;
  apellidos: string;
  rut: string;
  email: string;
  password: string;
  carrera: string;
  año_ingreso: number;
  rutLimpio?: string;
  carreraValida?: string;
  biografia?: string;
}

export interface Chat {
  _id: string;
  nombre: string;
  descripcion?: string;
  tipo: 'private' | 'group' | 'channel';
  participantes: User[];
  creador: User;
  fecha_creacion: Date;
  ultimo_mensaje?: Message;
  fecha_ultimo_mensaje?: Date;
  configuracion: {
    privacidad: 'public' | 'private';
    permite_invitaciones: boolean;
    max_participantes: number;
  };
}

export interface Message {
  id?: string;
  _id: string;
  chat?: string;
  remitente?: User;
  autor?: {
    id: string;
    name: string;
    foto_perfil?: string | null;
    carrera?: string;
  };
  mensaje?: string;
  contenido?: string;
  content?: string;
  tipo?: 'text' | 'image' | 'file' | 'audio';
  messageType?: string;
  fecha_envio?: Date | string;
  createdAt?: Date | string;
  leido?: boolean;
  is_read?: boolean;
  is_own?: boolean;
  editado?: boolean;
  fecha_edicion?: Date;
  respondiendo_a?: string;
}

export interface OnlineUser {
  userId: string;
  name: string;
  avatar?: string;
  connectedAt: Date;
}

export interface Post {
  _id: string;
  titulo: string;
  contenido: string;
  categoria: string;
  autor: User;
  likes: string[];
  comentarios: Comment[];
  fecha_creacion: Date;
  fecha_actualizacion?: Date;
  hasLiked?: boolean;
  tags?: string[];
  archivos?: string[];
}

export interface Comment {
  _id: string;
  post: string;
  autor: User;
  contenido: string;
  fecha_creacion: Date;
  fecha_actualizacion?: Date;
  likes: string[];
  respuestas?: Comment[];
  respondiendo_a?: string;
}

export interface SocketEvents {
  // Connection events
  user_online: (user: OnlineUser) => void;
  user_offline: (data: { userId: string; name: string; timestamp: Date }) => void;
  online_users: (users: OnlineUser[]) => void;
  
  // Chat events
  join_chat: (data: { chat_id: string }) => void;
  chat_joined: (data: { chat_id: string; message: string; timestamp: Date }) => void;
  user_joined_chat: (data: { userId: string; name: string; avatar?: string; chat_id: string; timestamp: Date }) => void;
  user_left_chat: (data: { userId: string; name: string; chat_id: string; timestamp: Date }) => void;
  
  // Message events
  send_message: (data: { chat_id: string; mensaje: string; tipo?: string }) => void;
  receive_message: (message: Message) => void;
  
  // Typing events
  typing: (data: { chat_id: string; isTyping: boolean }) => void;
  user_typing: (data: { userId: string; name: string; chat_id: string; isTyping: boolean; timestamp: Date }) => void;
  stop_typing: (data: { chat_id: string }) => void;
  
  // Error events
  error: (error: { type: string; message: string }) => void;
}

export interface ApiResponse<T = any> {
  success: boolean;
  message: string;
  data?: T;
  errors?: any[];
  errorsByField?: Record<string, string[]>;
}

// Carreras válidas UCN
export const VALID_CARRERAS = [
  'Ingeniería Civil',
  'Ingeniería Industrial', 
  'Ingeniería en Sistemas',
  'Ingeniería Comercial',
  'Arquitectura',
  'Medicina',
  'Enfermería',
  'Kinesiología',
  'Fonoaudiología',
  'Derecho',
  'Psicología',
  'Trabajo Social',
  'Administración Pública',
  'Contador Público',
  'Periodismo',
  'Publicidad',
  'Diseño Gráfico'
];

// Campus válidos UCN
export const VALID_CAMPUS = ['Antofagasta', 'Coquimbo', 'Santiago'];
